tortoise/tortoise.go

```diff
@@ -842,44 +846,52 @@ func (t *turtle) healingTortoise(ctx *tcontext, logger log.Log, lid types.LayerI
 
 // loops over all layers from the last verified up to a new target layer and attempts to verify each in turn.
 func (t *turtle) verifyLayers(ctx *tcontext) error {
+	target := t.Processed
 	logger := t.logger.WithContext(ctx).WithFields(
-		log.Named("verification_target", t.Last),
+		log.Named("verification_target", target),
 		log.Named("old_verified", t.Verified))
 
 	// attempt to verify each layer from the last verified up to one prior to the newly-arrived layer.
 	// this is the full range of unverified layers that we might possibly be able to verify at this point.
 	// Note: t.Verified is initialized to the effective genesis layer, so the first candidate layer here necessarily
 	// follows and is post-genesis. There's no need for an additional check here.
-	for candidateLayerID := t.Verified.Add(1); candidateLayerID.Before(t.Last); candidateLayerID = candidateLayerID.Add(1) {
-		logger := logger.WithFields(log.Named("candidate_layer", candidateLayerID))
+	for candlid := t.Verified.Add(1); candlid.Before(target); candlid = candlid.Add(1) {
+		logger := logger.WithFields(log.Named("candidate_layer", candlid))
 
 		// it's possible that self healing already verified a layer
-		if !t.Verified.Before(candidateLayerID) {
+		if !t.Verified.Before(candlid) {
 			logger.Info("self healing already verified this layer")
 			continue
 		}
 
 		logger.Info("attempting to verify candidate layer")
-		contextualValidity, err := t.verifyingTortoise(ctx, logger, candidateLayerID)
-		if err != nil {
-			return err
+		var (
+			contextualValidity map[types.BlockID]bool
+			err                error
+		)
+		// during rerun verifying tortoise prematurely verifies layers, which full tortoise won't be a
+		// able to verify because of additional eligible ballots with wrong beacon
+		if !(candlid.Before(t.layerCutoff()) && t.Last.Difference(candlid) > t.Zdist+t.ConfidenceParam) {
```

Comment:
i am not able to connect the comments and the code. why do we only run verifying tortoise for recent layers?

> don't use verifying tortoise for old layers, otherwise, it prematurely verifies layers because blocks with bad beacon are ignored. this problem will be resolved when ballot weight will be calculated properly.

isn't this by design? if there are two partitions with two different beacons, each partition should be happily verifying old layers until the rejoin. no?

Reply:
> isn't this by design? if there are two partitions with two different beacons, each partition should be happily verifying old layers until the rejoin. no?

yes, but i don't see any contradiction. it will work as you said. verifying tortoise should have the same output as full, but because bugs with weight it currently doesn't

we  cannot run verifying tortoise during rerun because atx weight in the implementation is meaningless (i am fixing it in next pr). so if after rejoin verifying tortoise will ignore bad blocks and erroneously verify layers which wouldn't be verified by full tortoise. because full tortoise will also count blocks with bad beacons.

Reply:
i think adding the if condition here is confusing. if this is a bug due to ATX weight, then maybe we should revisit this PR after the ATX weight is fixed.

in my mind, after rejoin, rerun will find out that there are a lot more weight ignored, and a different beacon. that should cause a revert in state because the old verified layers are no longer crossing the global threshold. self-healing/full tortoise should now kick in

Reply:
why is it confusing? in my opinion running full tortoise explicitly if we can is less confusing than using it as a fallback. but sure i can revert this change from this pr.

Reply:
> running full tortoise explicitly if we can is less confusing than using it as a fallback.

the confusing part for me is the special case with the old layers. my understanding is that the protocol, by design, avoid running full tortoise as much as possible. a fallback is exactly what it should be. self-healing should be triggered as rare as we possibly can. i think we should clarify with Tal about this too.

Reply:
i would prefer your ATX weight PR fix the bug first and re-evaluate the conditions that trigger self-healing/full tortoise.

Reply:
> the confusing part for me is the special case with the old layers

this is because there is a condition that doesn't allow full tortoise to run for new layers. 

> my understanding is that the protocol, by design, avoid running full tortoise as much as possible. a fallback is exactly what it should be.

there is no reason to do so, the consensus output of verifying and full should be the same by definition, and it is not much different in terms of computation. i will clarify but i don't see why it needs to be fallback

Reply:
@dshulyak this is because currently verifying tortoise is doing exactly the same thing as full tortoise, which is not the intention. verifying tortoise is supposed to only count margins only, not ALL the votes.